### PR TITLE
Adapt storage layer to new milestone ID based logic

### LINKED ISSUE #1418: Change REST API for milestone payloads
- [ ] Ensure iota.go/v3 hashes the milestone essence to generate milestone ID
- [ ] Return milestone payloads instead of messages
- [ ] Change `/milestones/{index}` to `/milestones/index/{index}`
- [ ] Add `/milestones/id/{milestoneId}`

### LINKED ISSUE #1421: Change database layout for Milestone Payloads
- [ ] Remove Index to MessageID Lookup
- [ ] Add Index to MilestoneID Lookup
- [ ] Add MilestoneID to MilestonePayload Lookup

### LINKED ISSUE #1434: Add `milestoneId` to milestones in `/info` endpoint

### PR DESCRIPTION
**Pull request by muXxer**

### CODE REVIEW

File: pkg/model/storage/message_metadata.go [hunk 1]
```diff
@@ -275,7 +258,6 @@ func (m *MessageMetadata) ObjectStorageValue() (data []byte) {
 	marshalUtil := marshalutil.New(23 + len(m.parents)*iotago.MessageIDLength)
```
Comment (alexsporn) [inline pkg/model/storage/message_metadata.go]: change to 19

File: pkg/testsuite/utils.go [hunk 2]
```diff
@@ -56,7 +56,7 @@ func (te *TestEnvironment) NewMessageBuilder(optionalTag ...string) *MessageBuil
 }
 
 func (b *MessageBuilder) LatestMilestonesAsParents() *MessageBuilder {
-	return b.Parents(hornet.MessageIDs{b.te.Milestones[len(b.te.Milestones)-1].Milestone().MessageID, b.te.Milestones[len(b.te.Milestones)-2].Milestone().MessageID})
+	return b.Parents(hornet.MessageIDs{b.te.coo.lastMilestoneMessageID})
```
Comment (alexsporn) [inline pkg/testsuite/utils.go:59]: this is now only the last one, before it was the last 2

Comment (alexsporn) [reply pkg/testsuite/utils.go:59]: Rename to ` LatestMilestoneAsParents()`

File: plugins/inx/server_milestones.go [hunk 3]
```diff
@@ -13,46 +13,26 @@ import (
...
 		Milestone: &inx.RawMilestone{
-			Data: bytes,
+			Data: ms.Milestone().Data(),
 		},
 	}, nil
 }
 
 func milestoneForIndex(msIndex milestone.Index) (*inx.Milestone, error) {
-	cachedMilestone := deps.Storage.CachedMilestoneOrNil(msIndex) // milestone +1
+	cachedMilestone := deps.Storage.CachedMilestoneByIndexOrNil(msIndex) // milestone +1
```
Comment (alexsporn) [inline plugins/inx/server_milestones.go:36]: Do you want to add in `ReadMilestone(...)` below a check, that if `GetMilestoneIndex() == 0`, that you look it up by `GetMilestoneId()` in the `MilestoneRequest`?

File: plugins/restapi/v2/node.go [hunk 4]
```diff
@@ -48,13 +58,27 @@ func info() (*infoResponse, error) {
...
+		},
+		Protocol: protocolResponse{
+			Version:       deps.ProtocolParameters.Version,
+			NetworkName:   deps.ProtocolParameters.NetworkName,
+			Bech32HRP:     string(deps.ProtocolParameters.Bech32HRP),
+			MinPowScore:   deps.ProtocolParameters.MinPowScore,
+			RentStructure: deps.ProtocolParameters.RentStructure,
+			TokenSupply:   iotago.EncodeUint64(deps.ProtocolParameters.TokenSupply),
 		},
-		Protocol: deps.ProtocolParameters,
```
Comment (alexsporn) [inline plugins/restapi/v2/node.go:57]: please revert this change, ProtocolParameters encode correctly to json.
Also remove `protocolResponse` type

File: plugins/restapi/v2/plugin.go [hunk 5]
```diff
@@ -60,9 +60,13 @@ const (
 	// MIMEVendorIOTASerializer => bytes
 	RouteTransactionsIncludedMessage = "/transactions/:" + restapipkg.ParameterTransactionID + "/included-message"
 
-	// RouteMilestone is the route for getting a milestone by its milestoneIndex.
+	// RouteMilestoneByIndex is the route for getting a milestone by its milestoneIndex.
 	// GET returns the milestone.
-	RouteMilestone = "/milestones/:" + restapipkg.ParameterMilestoneIndex
+	RouteMilestoneByIndex = "/milestones/index/:" + restapipkg.ParameterMilestoneIndex
```
Comment (alexsporn) [inline plugins/restapi/v2/plugin.go]: we said we wanted to use `by-index`

File: plugins/restapi/v2/plugin.go [hunk 6]
```diff
@@ -60,9 +60,13 @@ const (
...
+
+	// RouteMilestone is the route for getting a milestone by its ID.
+	// GET returns the milestone.
+	RouteMilestone = "/milestones/milestoneId/:" + restapipkg.ParameterMilestoneID
```
Comment (alexsporn) [inline plugins/restapi/v2/plugin.go]: we said we wanted to remove `milestoneId` from the path

File: plugins/restapi/v2/types.go [hunk 7]
```diff
@@ -10,15 +10,39 @@ import (
...
+	// The index of the milestone.
+	Index milestone.Index `json:"index"`
+	// The unix time of the milestone payload.
+	Timestamp uint32 `json:"timestamp"`
+	// The ID of the milestone.
+	MilestoneID string `json:"milestoneId"`
+}
+
+// protocolResponse defines the protocol info response.
+type protocolResponse struct {
```
Comment (alexsporn) [inline plugins/restapi/v2/types.go]: Remove this

File: plugins/restapi/v2/plugin.go (hunk 5 above)
Comment (muXxer) [reply plugins/restapi/v2/plugin.go]: oh, I just copied it from iota.go code, and this is based on our current TIP [swagger](https://editor.swagger.io/?url=https://raw.githubusercontent.com/iotaledger/tips/stardust-api/tips/TIP-0025/core-rest-api.yaml)

File: plugins/restapi/v2/plugin.go (hunk 6 above)
Comment (muXxer) [reply plugins/restapi/v2/plugin.go]: oh, I just copied it from iota.go code, and this is based on our current TIP [swagger](https://editor.swagger.io/?url=https://raw.githubusercontent.com/iotaledger/tips/stardust-api/tips/TIP-0025/core-rest-api.yaml)

File: plugins/restapi/v2/types.go [hunk 8]
```diff
@@ -40,10 +64,10 @@ type infoResponse struct {
 	Version string `json:"version"`
 	// The current status of this node.
 	Status nodeStatus `json:"status"`
+	// The protocol information of this node.
+	Protocol protocolResponse `json:"protocol"`
 	// The metrics of this node.
 	Metrics nodeMetrics `json:"metrics"`
-	// The protocol parameters used by this node.
-	Protocol *iotago.ProtocolParameters `json:"protocol"`
```
Comment (alexsporn) [inline plugins/restapi/v2/types.go:46]: Revert this

**Review by alexsporn — APPROVED** [review]
